### DISCUSSION TITLE
How do I write the date time to a Siemens PLC through the milo library?

### DISCUSSION THREAD

**Discussion by cool-wx** [eclipse-milo/milo#1354] [General]
How do I write the date time to a Siemens PLC through the milo library?

**Comment by kevinherron** [comment]
Do you know the NodeId of the Node you want to write to?

**Reply by cool-wx** [reply]
Yes, I know the NodeId, and I can read the time from the plc with this NodeId, but the type I read is a UByte[] array of length 8, and I parse the time from the UByte[] array, and the result is correct.However, when I write the time, I first try to write data of type Milo DateTime, prompt: “Bad_TypeMismatch“, then I try to write array of UByte[8] as well as parsing, and the result is successful, but the date time in plc is still wrong.

**Reply by kevinherron** [reply]
I believe with Siemens PLCs you must use the "LDT" datatype in the PLC in order for it to be an OPC UA DateTime in OPC UA.

**Reply by cool-wx** [reply]
I used the "LDT" data type in the PLC, and the time displayed in TIA Portal is: LDT#1990-01-01 00:00:00. My Siemens PLC is from the S7-1518.

**Reply by kevinherron** [reply]
LDT should show up as DateTime. If the write doesn't "take" that's not anything you can do on the Milo side. 

See https://support.industry.siemens.com/cs/mdm/59192925?c=170788318347&lc=en-MU for how Siemens maps S7 types to OPC UA.

**Reply by cool-wx** [reply]
Thank you, after I set it to DTL type (DT type at first), I solved the time writing problem by using the ExtendedObject encoder. Thank you very much!
